### PR TITLE
feat: Add Inclusion proofs for QGB verification

### PR DESCRIPTION
**Pull request by SweeXordious**
Adds support for the following proofs:
- Shares NMT proof to rows
- Rows merkle proof to data root
- data root merkle proof to commitment

Adds to: https://github.com/celestiaorg/celestia-app/issues/956
Closes: https://github.com/celestiaorg/celestia-core/issues/891

### CODE REVIEW

File: pkg/consts/consts.go
Comment (evan-forbes) [inline pkg/consts/consts.go:30]: I'm confused over the necessity to have 5 new queries. Want to discuss this synchronously?

File: proto/tendermint/types/types.proto
Comment (SweeXordious) [reply proto/tendermint/types/types.proto:240]: Yes, you can prove shares spanning multiple rows. What is not supported, is shares in multiple namespaces.

File: pkg/consts/consts.go
Comment (SweeXordious) [reply pkg/consts/consts.go:30]: Sure, as an overview (for other reviewers):
- `TxSharesQueryPath`: takes a transaction index and returns the shares range which the transaction takes. This is useful to get the rows proof.
- `MessageSharesQueryPath`: takes a transaction index and returns the shares range which the message, that the transaction committed to, occupies.
- `ShareInclusionProofQueryPath`: NMT inclusion proof of shares to rows
- `RowsInclusionProofQueryPath`: Merkle inclusion proof for rows to data root, defined by a range of rows.
- `RowsInclusionProofBySharesQueryPath`: merkle inclusion proof for rows to data root, defined by a range of shares. This one is useful, since we don't have access to the rows indexes in the shares NMT proof.

`RowsInclusionProofQueryPath` will not be used, I guess, for the QGB. But, I thought it is good to provide that API too, as it might be useful to just prove certain rows to data root.

Comment (SweeXordious) [reply pkg/consts/consts.go]: Check now please when you have time, refactored to 1 query containing everything

File: types/tx.go
Comment (evan-forbes) [inline types/tx.go:156]: [question]
don't feel like you have to revert if not, but was this related to this PR? like, was this necessary for something to work?

**Review by evan-forbes — COMMENTED** [review]
for the most part I think this looks good! :+1: . might want to rename something, but I need to think about it a bit more. would also like to see it work with the application before we merge

File: types/tx.go
Comment (SweeXordious) [reply types/tx.go:156]: Nope. It was just a cosmetic to be consistent.

**Comment by SweeXordious** [comment]
> would also like to see it work with the application before we merge

Yes 100%. That's why I'm leaving the PR as a draft.
I opened it to have first look on it, in case it needs any major change. 

Aside from that. Won't make it ready for review until I have everything plugged in nicely.

File: types/tx.go
Comment (SweeXordious) [reply types/tx.go]: [fb5aeaa](https://github.com/celestiaorg/celestia-core/pull/878/commits/fb5aeaa8d72d56e1bf407aa28f870f450461b466)

File: types/tx.go
Comment (evan-forbes) [inline types/tx.go:293]: [blocking docs]
lets leave a comment here or somewhere indidcating that this is tested in the application

**Review by evan-forbes — DISMISSED** [review]
I think we can merge, and we can even cut a release here if needed, but would prefer to keep using a commit in the app PR until we have a working integration test.

great work!! :+1:

**Comment by SweeXordious** [comment]
Integration test PR: https://github.com/celestiaorg/celestia-app/pull/1072

File: types/tx.go
Comment (SweeXordious) [inline types/tx.go:293]:
```suggestion
// Otherwise, it returns a sensible error.
// Note: these proofs are tested on the app side
```

File: light/rpc/client.go
Comment (rootulp) [inline light/rpc/client.go:468]: [question] is this a Merkle proof that proves the range of blocks [beginBlock:endBlock] exist in a Merkle tree where the root is data root? What Merkle tree is this? Does it exist in vanilla Tendermint?

File: proto/tendermint/types/types.proto
Comment (rootulp) [inline proto/tendermint/types/types.proto:224]:
```suggestion
  // data are the raw shares that are proven.
```

File: proto/tendermint/types/types.proto
Comment (rootulp) [inline proto/tendermint/types/types.proto:229]:
```suggestion
  // namespace_id is the namespaceID of the shares.
```

Comment (rootulp) [reply proto/tendermint/types/types.proto:229]: [question] is there a requirement that shares in this inclusion proof only belong to one namespace ID? Where is such a requirement enforced?

File: proto/tendermint/types/types.proto
Comment (rootulp) [inline proto/tendermint/types/types.proto:232]:
```suggestion
  // rows_proof are binary Merkle proofs of the rows containing the raw shares to the data root.
```

**Review by rootulp — COMMENTED** [review]
[question] I paused reviewing because I don't have enough context. Are there docs on the design / rationale for:

> Shares NMT proof to rows
> Rows merkle proof to data root
> data root merkle proof to commitment

File: light/rpc/client.go
Comment (SweeXordious) [reply light/rpc/client.go:468]: This is related to the QGB. 
The QGB contract receives commitments to a set of data roots, i.e. data commitments.  
This proof proves that a certain data root was part of the commitment (that was posted to QGB), which is defined by a range of blocks.

File: proto/tendermint/types/types.proto
Comment (SweeXordious) [reply proto/tendermint/types/types.proto:229]: it is an implementation limitation. creating inclusion proofs for multiple namespaces:

- more complex
- would require making changes to the `nmt` repository. Since it only supports proofs for a single namespace.

Also, from a QGB perspective, we will want to prove that some data was posted to a certain namespace. 

Ideally, it would be awesome to support any shares without constraints. But, it would take more time/effort to have it  working.

**Review by evan-forbes — APPROVED** [review]
I think we're good to merge, but ideally should triple check the integration test in the application

**Comment by SweeXordious** [comment]
@evan-forbes [61f09fd](https://github.com/celestiaorg/celestia-core/pull/878/commits/61f09fd5834ecb5e2fb1273127933de0fab8bc4e) check this please?

File: light/rpc/client.go
Comment (cmwaters) [reply light/rpc/client.go:468]: I am also a bit confused as to what the arguments `beginBlock` and `endBlock` mean. These are already loaded terms: https://github.com/celestiaorg/celestia-core/blob/a952ab34f0c2f691941c0c05729dcb34411ada7e/abci/types/application.go#L24-L26

Comment (cmwaters) [reply light/rpc/client.go:468]: Could we instead use something like `firstBlock` and `lastBlock`?

File: rpc/core/tx.go
Comment (cmwaters) [inline rpc/core/tx.go:204]: What happens if the value is nil and the logs are empty?

File: light/proxy/routes.go
Comment (cmwaters) [inline light/proxy/routes.go:32]: Is `prove_shares` supposed to be here? I notice you have the method in `light/rpc/client.go` but nothing ever calls that?

File: light/rpc/client.go
Comment (cmwaters) [inline light/rpc/client.go:475]: Currently, the light client is relying on whatever the full node responds with. Is there any way that the light client can verify this response themselves. Could they not fetch all the headers for the range of heights and build the inclusion proof themselves?

File: light/rpc/client.go
Comment (cmwaters) [inline light/rpc/client.go:516]: Same for this. The light client here is just trusting whatever the full node responds with. In some of the other methods like `Tx` the light client is actually able to prove that the transaction is committed as well as retrieve the transaction. Can this be done for this method?

**Review by cmwaters — COMMENTED** [review]
Left some general comments.

There is also an `openapi.yml` file that is used by Tendermint to document the RPC layer. Do we want to employ this as well for `celestia-core`? (cc @evan-forbes). If so, then I believe this PR should also update that file with the new methods

File: light/rpc/client.go
Comment (SweeXordious) [reply light/rpc/client.go]: [01cb80b](https://github.com/celestiaorg/celestia-core/pull/878/commits/01cb80bea07aef0f5da1e7457226e44b9fcd89fc)

File: rpc/core/blocks.go
Comment (SweeXordious) [reply rpc/core/blocks.go]: Yes, the `height` of the block we want to prove should be in the interval `[beginBlock, endBlock]`

File: rpc/core/tx.go
Comment (SweeXordious) [reply rpc/core/tx.go:204]: Good point, I don't have a strong reasoning around this. My conclusion was based on running the query a lot of times and seeing the outcome. What I noticed was that if a custom query fails, the value is nil and the logs contain some message

File: light/rpc/client.go
Comment (SweeXordious) [reply light/rpc/client.go]: that's true, do we want to support light clients generating the proofs ? @evan-forbes what do you think please?

File: light/proxy/routes.go
Comment (SweeXordious) [reply light/proxy/routes.go]: I guess there is a request for it:
```
func (c *baseRPCClient) DataRootInclusionProof(
	ctx context.Context,
	height uint64,
	firstBlock uint64,
	lastBlock uint64,
) (*ctypes.ResultDataRootInclusionProof, error) {
	result := new(ctypes.ResultDataRootInclusionProof)
	params := map[string]interface{}{
		"height":     height,
		"firstBlock": firstBlock,
		"lastBlock":  lastBlock,
	}

	_, err := c.caller.Call(ctx, "data_root_inclusion_proof", params, result)
	if err != nil {
		return nil, err
	}

	return result, nil
}
```

Does it need to be called somewhere else?

File: types/tx.go
Comment (SweeXordious) [inline types/tx.go:339]:
```suggestion
// It returns nil if the proof is valid. Otherwise, it returns a sensible error.
// The `root` is  the block data root that the shares to be proven belong to.
```

**Comment by SweeXordious** [comment]
> Left some general comments.
> 
> There is also an `openapi.yml` file that is used by Tendermint to document the RPC layer. Do we want to employ this as well for `celestia-core`? (cc @evan-forbes). If so, then I believe this PR should also update that file with the new methods

@evan-forbes  I guess that we should update the `openapi.yml` file, What do you think?

File: light/rpc/client.go
Comment (evan-forbes) [reply light/rpc/client.go]: yeah that makes a lot of sense, and is something we can plan for in celestia-node. Until then, nodes that need to produce proofs for the QGB will have to have access to a trusted celestia full node :grimacing:

**Comment by evan-forbes** [comment]
> There is also an openapi.yml file that is used by Tendermint to document the RPC layer. Do we want to employ this as well for celestia-core? (cc @evan-forbes). If so, then I believe this PR should also update that file with the new methods
@evan-forbes I guess that we should update the openapi.yml file, What do you think?

Yeah, that might be something that we need to do since we've added a few rpc endpoints. I'm also ok with doing this in a followup PR if we need to merge this one sooner rather than later

File: light/rpc/client.go
Comment (SweeXordious) [reply light/rpc/client.go]: https://github.com/celestiaorg/celestia-core/issues/924

File: light/rpc/client.go
Comment (SweeXordious) [reply light/rpc/client.go:516]: https://github.com/celestiaorg/celestia-core/issues/924

**Comment by SweeXordious** [comment]
Awesome, created an issue for this: https://github.com/celestiaorg/celestia-core/issues/925

**Review by evan-forbes — DISMISSED** [review]
all my feedback has been addressed so LGTM, and I'm approving!! :+1: 

I would defer to other reviewers before merging, however.

**Comment by evan-forbes** [comment]
If there's no further feedback by end of Monday @SweeXordious, provided we have issues for everything that we still need to do, I think its okay to merge

**Comment by evan-forbes** [comment]
can merge?

**Comment by SweeXordious** [comment]
I resolve the conflicts and :shipit:

**Comment by SweeXordious** [comment]
Just ran a round trip test:
- Send a transaction to Celestia
- commit to the attestations on the QGB
- check if the transaction and the msg were attested to
And, it ran successfuly :tada: 
Will merge once the CI is green

**Pull request merged**
